src/server/list_family.cc

```diff
@@ -128,6 +128,33 @@ bool ElemCompare(const quicklistEntry& entry, string_view elem) {
   return elem == an.Piece();
 }
 
+// Circular buffer for string messages
+struct CircularMessages {
+  CircularMessages(size_t size) : current{0}, messages{size} {
+  }
+
+  void operator<<(string msg) {
+    messages[current] = msg;
+    current = (current + 1) % messages.size();
+  }
+
+  vector<string> All() {
+    vector<string> out;
+    out.insert(out.end(), messages.begin() + current, messages.end());
```

Comment:
Nit: instead of `erase()`ing below, you could `copy_if()` in the first pass. The second path, I think, should be ok to be plainly copied as it could never have empty values.